bin/install_windows_10_sdk.ps1

```
@@ -0,0 +1,59 @@
+# Stop script execution when a non-terminating error occurs
+$ErrorActionPreference = "Stop"
+
+Write-Host "--- :windows: Installing Windows 10 SDK and Visual Studio Build Tools"
+
+$windowsSDKVersionFile = ".windows-10-sdk-version"
+if (-not (Test-Path $windowsSDKVersionFile)) {
+  Write-Output "[!] No Windows 10 SDK version file found at $windowsSDKVersionFile."
+  exit 1
+}
+
+$windows10SDKVersion = Get-Content $windowsSDKVersionFile
+
+Write-Host "Will attempt to set up Windows 10 ($windows10SDKVersion) SDK and Visual Studio Build Tools..."
+
+# Download the Visual Studio Build Tools Bootstrapper
+Write-Output "~~~ Downloading Visual Studio Build Tools..."
+
+$buildToolsPath = ".\vs_buildtools.exe"
+
+Invoke-WebRequest `
+  -Uri https://aka.ms/vs/17/release/vs_buildtools.exe `
+  -OutFile $buildToolsPath
+
+If (-not (Test-Path $buildToolsPath)) {
+  Write-Output "[!] Failed to download Visual Studio Build Tools"
+  Exit 1
+} else {
+  Write-Output "Successfully downloaded Visual Studio Build Toosl at $buildToolsPath."
+}
+
+# Install the Windows SDK and other required components
+Write-Output "~~~ Installing Visual Studio Build Tools..."
+Start-Process `
+  -FilePath $buildToolsPath `
+  -ArgumentList "--quiet --wait --add Microsoft.VisualStudio.Component.Windows10SDK.$windows10SDKVersion" `
+  -NoNewWindow `
+  -Wait
+
+# Check if the installation was successful in file system
+$windowsSDKsRoot = "C:\Program Files (x86)\Windows Kits\10\bin"
+$sdkPath = "$windowsSDKsRoot\10.0.$windows10SDKVersion.0\x64"
+If (-not (Test-Path $sdkPath)) {
+  Write-Output "[!] Failed to install Windows 10 SDK: Could not find SDK at $sdkPath."
+  If (-not (Test-Path $windowsSDKsRoot)) {
+    Write-Output "[!] Expected $windowsSDKsRoot to exist, but it does not."
+  } else {
+    Write-Output "    Found:"
+    Get-ChildItem -Path $windowsSDKsRoot | ForEach-Object { Write-Output "    - $windowsSDKsRoot\$_" }
+  }
+  Exit 1
+}
```

Is there an easy way to check if `$windows10SDKVersion` is valid (i.e. an existing version, as opposed to if someone tried to use `latest` or `none` or whatever invalid text in the version file) and detect it early?

In particular, I'm wondering if the `vs_buildtools.exe` would fail with an explicit and clear error if that `-add …` argument we pass to it was invalid (i.e. if it didn't find the component), allowing us to print a specific error message in that case (like "Check the value you set in `$windowsSDKVersionFile`"), as opposed to the installer happily continuing with the installation of all the other default components—without complaining about this one not existing in the options—and us only finding out the component was not installed **_after the fact_**, thanks to your test on lines 42–51… (and even in that case, would it help to provide in the error messages a more explicit suggestion to check the syntax/version used in the version file?)

---

>  I'm wondering if the `vs_buildtools.exe` would fail with an explicit and clear error

I've done a few experiments in my VM. It fails silently, with exit code `1`.

---

I'll duplicated what I wrote in https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/153#discussion_r1962609092

> This all makes me wonder whether we might be better off using a default SDK version in the installer, and only falling back to the version file and/or a parameter if the user provides one. After all, I implemented the version file because other tools use it and because the app I was working on had a check for the SDK being installed and so I wanted a single source of truth. But I'd argue now that our tooling takes care of the installing, that check is redundant: the build script should trust the SDK is available.

I think the conversation should continue there, given that's the PR that is editing the script.

I also like the idea of having a list of valid numbers to pick from, given it's published at https://learn.microsoft.com/en-us/visualstudio/install/workload-component-id-vs-build-tools?view=vs-2022 but I think that would be a good additional option, while we should still default to installing the latest valid version for the user ourselves.

---

Maybe if we end up implementing the behavior of defaulting to install the latest valid version, we could allow `latest` to be a special valid content for `.windows-10-sdk-version`, that way if the file is not present it won't install the SDK at all, but if it's present with `latest` it will install the latest (and if it's present with a specific version it will deterministically install that specific version, acting as a lockfile) 🤔 

That way the implicit auto-install from `prepare_windows_host_for_app_distibution` if file is present would still work for that default-version fallback… 🤔